src/diff_generate.c

```diff
@@ -694,15 +696,37 @@ static int maybe_modified_submodule(
 	git_submodule *sub;
 	unsigned int sm_status = 0;
 	git_submodule_ignore_t ign = diff->base.opts.ignore_submodules;
+	git_strmap *submodule_cache = NULL;
 
 	*status = GIT_DELTA_UNMODIFIED;
 
 	if (DIFF_FLAG_IS_SET(diff, GIT_DIFF_IGNORE_SUBMODULES) ||
 		ign == GIT_SUBMODULE_IGNORE_ALL)
 		return 0;
 
-	if ((error = git_submodule_lookup(
-			&sub, diff->base.repo, info->nitem->path)) < 0) {
+	if (diff->base.repo->submodule_cache != NULL) {
+		submodule_cache = diff->base.repo->submodule_cache;
+	} else {
+		if (!info->submodule_cache_initialized) {
+			info->submodule_cache_initialized = true;
```

Comment:
Can we maybe pull this into a helper function?  `git_submodule_cache_init(git_strmap **out)`?  It would actually dovetail nicely with a `git_submodule_cache_free` that `git_diff__from_iterators` could call.

If you really wanted to be fancy, we could create a `git_submodule_cache` struct that has a refcount so that you could create a new one or borrow the repository's and we could drop the `submodule_cache_initialized` out of the diff struct.  But this is probably more work for limited gain.

Reply:
sounds good!

i ended up not adding the refcount (for now) since there are some cases where the upfront cache generation fails (see https://github.com/libgit2/libgit2/runs/1535037963?check_suite_focus=true#step:7:1289 for an example), but looking up the submodules individually succeeds. so i still want to have that explicit `submodule_cache_initialized` flag to allow for this case.